pkg/netlight/resource/resource.go

```diff
@@ -0,0 +1,399 @@
+	}
+
+	if err := nft.Apply(rules, nsName); err != nil {
+		return nil, fmt.Errorf("failed to apply nft rules for namespace '%s': %w", name, err)
+	}
+
+	return &Resource{name}, setupMycelium(netNS, infMycelium, seed)
+}
+
+func Delete(name string) error {
+	nsName := fmt.Sprintf("n%s", name)
+	netNS, err := namespace.GetByName(nsName)
+	if errors.Is(err, os.ErrNotExist) {
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if err := destroyMycelium(netNS, zinit.Default()); err != nil {
+		return err
+	}
+
+	if err := namespace.Delete(netNS); err != nil {
+		return err
+	}
+
+	privateNetBr := fmt.Sprintf("r%s", name)
+	myBr := fmt.Sprintf("m%s", name)
+
+	if err := bridge.Delete(privateNetBr); err != nil {
+		return err
+	}
+
+	return bridge.Delete(myBr)
+
+}
+
+func setLinkAddr(name string, ip *net.IPNet) error {
+	link, err := netlink.LinkByName(name)
+	if err != nil {
+		return fmt.Errorf("failed to set link address: %w", err)
+	}
+
+	// if err := options.Set(name, options.IPv6Disable(false)); err != nil {
+	// 	return fmt.Errorf("failed to enable ip6 on interface %s: %w", name, err)
+	// }
+
+	addr := netlink.Addr{
+		IPNet: ip,
+	}
+	err = netlink.AddrAdd(link, &addr)
+	if err != nil && !os.IsExist(err) {
+		return fmt.Errorf("failed to add ip address to link: %w", err)
+	}
+
+	return netlink.LinkSetUp(link)
+}
+
+// Get return resource handler
+func Get(name string) (*Resource, error) {
+	nsName := fmt.Sprintf("n%s", name)
+
+	if namespace.Exists(nsName) {
+		return &Resource{name}, nil
+	}
+
+	return nil, fmt.Errorf("resource not found: %s", name)
+}
+
+var networkResourceNet = net.IPNet{
+	IP:   net.ParseIP("100.64.0.0"),
+	Mask: net.IPv4Mask(0xff, 0xff, 0, 0),
+}
+
+func (r *Resource) AttachPrivate(id string, vmIp net.IP) (device pkg.TapDevice, err error) {
+	nsName := fmt.Sprintf("n%s", r.name)
+	netNs, err := namespace.GetByName(nsName)
+	if err != nil {
+		return
+	}
+	if vmIp[3] == 1 {
+		return device, fmt.Errorf("ip %s is reserved", vmIp.String())
+	}
+
+	var addrs []netlink.Addr
+
+	if err = netNs.Do(func(_ ns.NetNS) error {
+		link, err := netlink.LinkByName(infPrivate)
+		if err != nil {
+			return err
+		}
+		addrs, err = netlink.AddrList(link, netlink.FAMILY_V4)
+		if err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		return
+	}
+	if len(addrs) != 1 {
+		return device, fmt.Errorf("expect addresses on private interface to be 1 got %d", len(addrs))
+	}
+	gw := addrs[0].IPNet
+	if !gw.Contains(vmIp) {
+		return device, fmt.Errorf("ip not in range")
+	}
+
+	deviceName := ifaceutil.DeviceNameFromInputBytes([]byte(id))
+	tapName := fmt.Sprintf("b-%s", deviceName)
+
+	privateNetBr := fmt.Sprintf("r%s", r.name)
+	hw := ifaceutil.HardwareAddrFromInputBytes([]byte(tapName))
+
+	ip := &net.IPNet{
+		IP:   vmIp,
+		Mask: gw.Mask,
+	}
+	_, getLinkErr := netlink.LinkByName(tapName)
+	if getLinkErr != nil {
+		mtap, err := macvtap.CreateMACvTap(tapName, privateNetBr, hw)
```

Comment:
We create TAP devices for VMs not Macvtap. this makes it un0necessary complex to pass this device to the VM when it can just be a tap device